src/Psalm/Internal/Provider/ReturnTypeProvider/TriggerErrorReturnTypeProvider.php

```diff
@@ -0,0 +1,67 @@
+<?php declare(strict_types=1);
+
+namespace Psalm\Internal\Provider\ReturnTypeProvider;
+
+use Psalm\Plugin\EventHandler\Event\FunctionReturnTypeProviderEvent;
+use Psalm\Type;
+
+use function in_array;
+
+use const E_USER_DEPRECATED;
+use const E_USER_ERROR;
+use const E_USER_NOTICE;
+use const E_USER_WARNING;
+
+class TriggerErrorReturnTypeProvider implements \Psalm\Plugin\EventHandler\FunctionReturnTypeProviderInterface
+{
+    /**
+     * @return array<lowercase-string>
+     */
+    public static function getFunctionIds(): array
+    {
+        return ['trigger_error'];
+    }
+
+    public static function getFunctionReturnType(FunctionReturnTypeProviderEvent $event): ?Type\Union
+    {
+        $config = $event->getStatementsSource()->getCodebase()->config;
+        if ($config->trigger_error_behaviour === 'always') {
+            return new Type\Union([new Type\Atomic\TNever()]);
+        }
+
+        if ($config->trigger_error_behaviour === 'none') {
+            return new Type\Union([new Type\Atomic\TTrue()]);
+        }
+
+        //default behaviour
+        $call_args = $event->getCallArgs();
+        $statements_source = $event->getStatementsSource();
+        if (isset($call_args[1])
+            && ($array_arg_type = $statements_source->getNodeTypeProvider()->getType($call_args[1]->value))
+        ) {
+            $return_types = [];
+            foreach ($array_arg_type->getAtomicTypes() as $atomicType) {
+                if ($atomicType instanceof Type\Atomic\TLiteralInt) {
+                    if (in_array($atomicType->value, [E_USER_WARNING, E_USER_DEPRECATED, E_USER_NOTICE], true)) {
+                        $return_types[] = new Type\Atomic\TTrue();
+                    } elseif ($atomicType->value === E_USER_ERROR) {
+                        $return_types[] = new Type\Atomic\TNever();
+                    } else {
+                        // not recognized int literal. return false before PHP8, fatal error since
+                        $return_types[] = new Type\Atomic\TFalse();
+                    }
+                } elseif ($atomicType instanceof Type\Atomic\TInt) {
+                    // not specific integer. Returning a mix of never and bool would not make sense for Psalm
+                    $return_types[] = new Type\Atomic\TMixed();
```

Comment:
So that should be `bool` then.

Reply:
I started with bool|never but it lead to Psalm reporting an error when trying to return this type. Bool however seem to narrow, given that never is a possibility, don't you think?

Reply:
`never` is a bottom type, so `bool|never` is not wider than `bool`, but is identical to it.